src/util/http/websocket/EphemeralWaitingList.cpp

```diff
@@ -0,0 +1,68 @@
+//   Copyright 2023, University of Freiburg,
+//   Chair of Algorithms and Data Structures.
+//   Author: Robin Textor-Falconi <textorr@informatik.uni-freiburg.de>
+
+#include "util/http/websocket/EphemeralWaitingList.h"
+
+#include <boost/asio/associated_cancellation_slot.hpp>
+#include <boost/asio/dispatch.hpp>
+
+namespace ad_utility::websocket {
+
+void EphemeralWaitingList::signalQueryStart(const QueryId& queryId) {
+  auto [start, stop] = waitingCallbacks_.equal_range(queryId);
+  for (auto it = start; it != stop; ++it) {
+    auto& [func, id] = it->second;
+    func();
+  }
+  waitingCallbacks_.erase(queryId);
+}
+
+template <typename CompletionToken>
+auto EphemeralWaitingList::registerCallbackAndWait(
+    const QueryId& queryId, CompletionToken&& token) {
+  auto initiate = [this, &queryId]<typename Handler>(Handler&& self) mutable {
+    auto sharedSelf = std::make_shared<Handler>(std::forward<Handler>(self));
+    auto cancellationSlot = net::get_associated_cancellation_slot(
+        *sharedSelf, net::cancellation_slot());
+    FunctionId functionId{idCounter_++};
+    auto executor = net::get_associated_executor(*sharedSelf);
+    waitingCallbacks_.emplace(
+        queryId, IdentifiableFunction{[sharedSelf = std::move(sharedSelf)]() {
+                                        (*sharedSelf)();
+                                      },
+                                      functionId});
+
+    if (cancellationSlot.is_connected()) {
+      cancellationSlot.assign(
+          [this, &queryId, &functionId,
+           executor = std::move(executor)](net::cancellation_type) {
+            // Make sure to remove the entry on the same executor
```

Comment:
Is it really safe to pass in the `queryId` and the `functionId` by reference into that lambda that might be called who knows when. (I am pretty sure that at least the `functionId` is definitely hanging, and the `queryId` is rather dangerous.

Reply:
The queryId's lifetime in this case is tied to the lifetime of the coroutine handling the websocket request, so it's definitely still alive while said coroutine is still in the process of being cancelled, or rather a sub-awaitable within the call stack of said coroutine (which can catch the cancellation and proceed normally)